### PR TITLE
Link FrameworkList.xml to a place where MSBuild SDK actually expects it

### PR DESCRIPTION
**Pull request by filipnavara**
The [SDK conflict resolution code](https://github.com/dotnet/sdk/blob/3c1307d22747bac8c499a711d1bac39d7fdba541/src/Tasks/Common/src/ConflictResolution/ResolvePackageFileConflicts.cs#L54-L57) expects the FrameworkList.xml file to be located in `$(TargetFrameworkDirectory)/RedistList/FrameworkList.xml`.

Should fix
https://github.com/dotnet/runtime/issues/49211
https://github.com/dotnet/runtime/issues/49940
https://github.com/dotnet/runtime/issues/49477
https://github.com/xamarin/xamarin-macios/issues/10912
https://github.com/xamarin/xamarin-macios/issues/10839
https://github.com/xamarin/xamarin-macios/issues/10548
https://github.com/xamarin/xamarin-macios/issues/10592
https://github.com/mono/mono/issues/20805
https://github.com/mono/mono/issues/20894

### CODE REVIEW

**Comment by rolfbjarne** [comment]
This is the failure:

```
ln: /Users/builder/azdo/_work/1/s/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/lib/mono/Xamarin.iOS/RedistList/FrameworkList.xml: No such file or directory
make[1]: *** [/Users/builder/azdo/_work/1/s/xamarin-macios/_ios-build/Library/Frameworks/Xamarin.iOS.framework/Versions/git/lib/mono/Xamarin.iOS/RedistList/FrameworkList.xml] Error 1
make[1]: *** Waiting for unfinished jobs....
```

My guess is that the target directory doesn't exist.

You can probably repro this locally by doing this from the root directory (this will do a clean build):

```
make git-clean-all
make all -j8
```

**Comment by filipnavara** [comment]
Thanks. I am doing the local build on clean machine now. I likely didn't notice it because I was testing the steps manually before and I created the directories and links by hand earlier.

**Review by mandel-macaque — APPROVED** [review]

**Review by spouliot — COMMENTED** [review]
👍

**Review by rolfbjarne — APPROVED** [review]

**Comment by rolfbjarne** [comment]
I don't think we've had a PR that fixes that many issues before..

**Review by dalexsoto — APPROVED** [review]

**Comment by filipnavara** [comment]
I'd like to at least manually verify the packages once they are build. I also have a build test written locally but I need to fix my Xcode installation first so I can test it.

**Review by chamons — APPROVED** [review]
🎉

**Comment by mandel-macaque** [comment]
@filipnavara I'll make sure that pkgs are created and will post links here.

**Comment by mandel-macaque** [comment]
@filipnavara  FYI: https://github.com/xamarin/xamarin-macios/commit/ba7c8370d15dd3413cee235b5068cc7edf85d763

**Comment by filipnavara** [comment]
@mandel-macaque Thanks a lot.

I'm still playing with some trivial test to ensure that everything is working properly: 
https://github.com/xamarin/xamarin-macios/commit/4af5fd513d886cf9197957bf8b503fbfc963d683

**Comment by mandel-macaque** [comment]
@filipnavara do you need that last commit too?

**Comment by filipnavara** [comment]
@mandel-macaque Nah, just the first two.

**Comment by filipnavara** [comment]
Unfortunately the packaging fails [which I anticipated] by having the symlinks point to wrong location:
```
filipnavara@172-4-1-17 app % ls -l /Library/Frameworks/Xamarin.Mac.framework/Versions/Current/lib/mono/Xamarin.Mac/RedistList/
total 0
lrwxr-xr-x  1 root  wheel  161 Mar 23 07:57 FrameworkList.xml -> /Users/builder/azdo/_work/1/s/maccore/release/_mac-install/Library/Frameworks/Xamarin.Mac.framework/Versions/7.11.0.225/lib/mono/4.5/RedistList/FrameworkList.xml
```

Is there any way to make the packages locally? Or can you give me pointer on how the other symlinks are handled? I also have an alternative solution without symlinks but I would like to understand the process nevertheless.

**Comment by rolfbjarne** [comment]
> Unfortunately the packaging fails [which I anticipated] by having the symlinks point to wrong location:
> 
> ```
> filipnavara@172-4-1-17 app % ls -l /Library/Frameworks/Xamarin.Mac.framework/Versions/Current/lib/mono/Xamarin.Mac/RedistList/
> total 0
> lrwxr-xr-x  1 root  wheel  161 Mar 23 07:57 FrameworkList.xml -> /Users/builder/azdo/_work/1/s/maccore/release/_mac-install/Library/Frameworks/Xamarin.Mac.framework/Versions/7.11.0.225/lib/mono/4.5/RedistList/FrameworkList.xml
> ```
> 
> I suppose I need to modify it to make the symlink with absolute path.

This is handled by using $(IOS_TARGETDIR) instead of $(IOS_DESTDIR): the former is empty when building a package, and the same as IOS_DESTDIR when building locally.

Example:

https://github.com/xamarin/xamarin-macios/blob/b2248993de13ab31d3c66b74820cb3f0077032c6/msbuild/Makefile#L247-L248

> Is there any way to make the packages locally? Or can you give me pointer on how the other symlinks are handled? I also have an alternative solution without symlinks but I would like to understand the process nevertheless.

Unfortunately the code to build the package is not public.

**Comment by filipnavara** [comment]
@rolfbjarne Thanks. Makes a lot more sense. Ready for take two on the package build :-)

**Comment by rolfbjarne** [comment]
New build has been scheduled!

**Comment by rolfbjarne** [comment]
Failure:

```
[...]
[BUILD] mlaunch
/Library/Frameworks/Mono.framework/Versions/Current/lib/mono/msbuild/15.0/bin/Microsoft.Common.CurrentVersion.targets(84,5): error MSB4184: The expression "[Microsoft.Build.Utilities.ToolLocationHelper]::GetPathToStandardLibraries(Xamarin.Mac, v2.0, '', '', '', /Users/builder/azdo/_work/1/s/xamarin-macios/_mac-build/Library/Frameworks/Mono.framework/External/xbuild-frameworks;/Library/Frameworks/Mono.framework/External/xbuild-frameworks)" cannot be evaluated. There was an error reading the redist list file "/Users/builder/azdo/_work/1/s/xamarin-macios/_mac-build/Library/Frameworks/Xamarin.Mac.framework/Versions/Current/lib/mono/Xamarin.Mac/RedistList/FrameworkList.xml". Could not find file "/Users/builder/azdo/_work/1/s/xamarin-macios/_mac-build/Library/Frameworks/Xamarin.Mac.framework/Versions/Current/lib/mono/Xamarin.Mac/RedistList/FrameworkList.xml" [/Users/builder/azdo/_work/1/s/maccore/tools/mlaunch/Xamarin.Hosting/Xamarin.Hosting.csproj]
/Library/Frameworks/Mono.framework/Versions/Current/lib/mono/msbuild/15.0/bin/Microsoft.Common.CurrentVersion.targets(84,5): error MSB4184: The expression "[Microsoft.Build.Utilities.ToolLocationHelper]::GetPathToStandardLibraries(Xamarin.Mac, v2.0, '', '', '', /Users/builder/azdo/_work/1/s/xamarin-macios/_mac-build/Library/Frameworks/Mono.framework/External/xbuild-frameworks;/Library/Frameworks/Mono.framework/External/xbuild-frameworks)" cannot be evaluated. There was an error reading the redist list file "/Users/builder/azdo/_work/1/s/xamarin-macios/_mac-build/Library/Frameworks/Xamarin.Mac.framework/Versions/Current/lib/mono/Xamarin.Mac/RedistList/FrameworkList.xml". Could not find file "/Users/builder/azdo/_work/1/s/xamarin-macios/_mac-build/Library/Frameworks/Xamarin.Mac.framework/Versions/Current/lib/mono/Xamarin.Mac/RedistList/FrameworkList.xml" [/Users/builder/azdo/_work/1/s/maccore/tools/mlaunch/Xamarin.Hosting/Xamarin.Launcher/Xamarin.Launcher.csproj]
/Library/Frameworks/Mono.framework/Versions/Current/lib/mono/msbuild/15.0/bin/Microsoft.Common.CurrentVersion.targets(84,5): error MSB4184: The expression "[Microsoft.Build.Utilities.ToolLocationHelper]::GetPathToStandardLibraries(Xamarin.Mac, v2.0, '', '', '', /Users/builder/azdo/_work/1/s/xamarin-macios/_mac-build/Library/Frameworks/Mono.framework/External/xbuild-frameworks;/Library/Frameworks/Mono.framework/External/xbuild-frameworks)" cannot be evaluated. There was an error reading the redist list file "/Users/builder/azdo/_work/1/s/xamarin-macios/_mac-build/Library/Frameworks/Xamarin.Mac.framework/Versions/Current/lib/mono/Xamarin.Mac/RedistList/FrameworkList.xml". Could not find file "/Users/builder/azdo/_work/1/s/xamarin-macios/_mac-build/Library/Frameworks/Xamarin.Mac.framework/Versions/Current/lib/mono/Xamarin.Mac/RedistList/FrameworkList.xml" [/Users/builder/azdo/_work/1/s/maccore/tools/mlaunch/Xamarin.Hosting/Xamarin.Localization.Mlaunch/Xamarin.Localization.Mlaunch.csproj]
```

This is in a private part of our build, so I'll have a look.

**Comment by filipnavara** [comment]
> This is in a private part of our build, so I'll have a look.

It's the symlinks pointing to non-existing file. Just fails at that particular point because it's the first task accessing the file through the symlink.

Sorry for the slow turn around, it's quite tricky for me to test these changes locally.

**Comment by filipnavara** [comment]
I manually verified the latest package on my test case so dropping the draft status.

**Comment by dalexsoto** [comment]
We should backport this to xcode12.5 and d16-9 once it is merged

**Comment by filipnavara** [comment]
There's still some Mac Catalyst test failing. I'll try to see if I can reproduce it locally with the packages. If there's any more detailed error message it would be helpful.

**Comment by dalexsoto** [comment]
@filipnavara I think your changes are ok and unrelated to the failure on device tests, the PR tests seems to be fine https://github.com/xamarin/xamarin-macios/pull/10928#commitcomment-48619930 I'll let @rolfbjarne confirm and the merge decision.

**Comment by rolfbjarne** [comment]
The only test result we care about at the moment is this one: https://github.com/xamarin/xamarin-macios/pull/10928#commitcomment-48619930, and that's green.

**Review by rolfbjarne — APPROVED** [review]

**Pull request merged**
